File: solutions/other/graph3p1.py
n, m = map(int, input().split())
adj = [[] for _ in range(30005)] # the neighbours of i
for i in range(m):
    x, y = map(int, input().split())
    adj[x].append(y) # y is a neighbour of x
    adj[y].append(x) # x is a neighbour of y

q = []
vis = [False for _ in range(30005)]
dis = [-1 for _ in range(30005)]

# getting k sources
k = int(input())
for i in range(k):
    source = int(input())
    q.append(source)
    vis[source] = True
    dis[source] = 0

while len(q) > 0:
    cur = q.pop(0)
    # process neighbours
    for x in adj[cur]:
        if not vis[x]:
            vis[x] = True
            dis[x] = dis[cur] + 1
            q.append(x)

print(max(dis))

File: solutions/other/graph3p1.cpp
#include <iostream>
#include <string>
#include <vector>
#include <queue>

using namespace std;

int n, m, k;
vector<int> adj[30005];
bool vis[30005];
int dis[30005];

int main() {
    cin >> n >> m;
    for (int i = 0; i < m; i++) {
        int x, y;
        cin >> x >> y;
        adj[x].push_back(y);
        adj[y].push_back(x);
    }
    queue<int> q;

    cin >> k;
    for (int i = 0; i < k; i++) {
        int x; cin >> x;
        q.push(x);
        vis[x] = true;
    }
    while (!q.empty()) {
        int cur = q.front();
        q.pop();
        for (int x : adj[cur]) {
            if (!vis[x]) {
                vis[x] = true;
                q.push(x);
                dis[x] = dis[cur] + 1;
            }
        }
    }
    int ans = 0;
    for (int i = 0; i < 30005;i++) {
        ans = max(ans, dis[i]);
    }
    cout << ans << "\n";
}